ibtk/include/ibtk/PartitioningBox.h

```diff
@@ -0,0 +1,146 @@
+ * defined by two points. Unlike a standard bounding box, a PartitioningBox is
+ * an <code>NDIM</code>-dimensional tensor product of half-open intervals:
+ * i.e., it is a half-open box. This property allows one to partition a domain
+ * into a set of boxes.
+ *
+ * It is possible for a partitioning box to be 'empty': in that case the
+ * bounds are [oo, oo) in each dimension.
+ */
+class PartitioningBox
+{
+public:
+    /// Default Constructor: an 'empty' partitioning box.
+    PartitioningBox();
+
+    /// Constructor.
+    PartitioningBox(const Point &bottom_point,
+                    const Point &top_point);
+
+    /// Constructor, starting from a SAMRAI data type.
+    PartitioningBox(const SAMRAI::geom::CartesianPatchGeometry<NDIM> &patch);
+
+    /// Get the bottom left corner of the box.
+    const Point &bottom() const;
+
+    /// Get the top right corner of the box. Note that this point is not in
+    /// the box.
+    const Point &top() const;
+
+    /// Return true if the point is inside the box and false otherwise.
+    bool contains(const Point &point) const;
+
+    /// Return the volume of the box. If the box is empty then this is zero.
+    double volume() const;
+
+protected:
+    /// bottom left and top right corners of the bounding box.
+    std::pair<Point, Point> d_bounding_points;
+};
+
+/*!
+ * \brief Class PartitioningBoxes stores a set of bounding boxes and can check
+ * if a point is in the set of partitioning boxes or not in a more optimized
+ * way than just looping over a std::vector<PartitioningBox>.
+ *
+ * It is possible for the set of bounding boxes to be empty: in that case the
+ * bounds are [oo, oo) in each dimension.
+ */
+class PartitioningBoxes
+{
+public:
+    /// Default constructor: an empty collection of partitioning boxes with
+    /// bounds at oo.
+    PartitioningBoxes() = default;
+
+    /// Constructor. Creates an object from a sequence of partitioning boxes.
+    template <typename ForwardIterator>
+    PartitioningBoxes(const ForwardIterator begin, const ForwardIterator end);
+
+    /// Constructor. Uses the finest level boxes on the provided
+    /// PatchHierarchy to create a set of bounding boxes.
+    PartitioningBoxes(const SAMRAI::hier::PatchHierarchy<NDIM> &hierarchy);
+
+    /// Get the bottom left corner of the partitioning box bounding all other
+    /// partitioning boxes.
+    const Point &bottom() const;
+
+    /// Get the top right corner of the partitioning box bounding all other
+    /// partitioning boxes. Note that this point is not in the collection of
+    /// boxes.
+    const Point &top() const;
+
+    /// Return true if the point is inside the set of boxes and false
+    /// otherwise.
+    bool contains(const Point &point) const;
+
+    /// Return a pointer to the first partitioning box.
+    const PartitioningBox * begin() const;
+
+    /// Return a pointer to one past the end of the end of the partitioning
+    /// box array.
+    const PartitioningBox * end() const;
```

Comment:
Worth providing a range for range-based for?

Reply:
Unless I am mistaken providing the pair there is enough to get range-based for working. I did use one laster.

Reply:
OK great! You know better than me.